### ISSUE TITLE
Psalm incorrectly unions closures with different param types

### ISSUE THREAD

**Issue by Akhristenko** [vimeo/psalm#2157]
let we have class with property that contain array of callable(int):int.
And we have method which use this array.
In one method we add to this array callable(bool):int
Psalm don't show any errors, instead he change type of property, but only for method where we add callable(bool):int
As a result, this leads to an error in runtime.
Example: [https://psalm.dev/r/23cebb6207](https://psalm.dev/r/23cebb6207)

**Comment by muglug** [comment]
Thanks, this is really a bigger issue:

```php
class A {}
class B {}

function foo() : void {
  $a = [
    function(A $f):void {},
    function(B $i):void {},
  ];

  foreach ($a as $c) {
    $c(new A());
  }
}
```

Expected: `PossiblyInvalidArgument`
Actual: No issue
